File: src/build_utils.py
# ==============================================================
#						 build_utils.py
# ==============================================================
import os, zipfile
import xml.etree.ElementTree as ET
import string
import secrets

from variables import *
from core_utils import *

""" ------------------ Functions ------------------
  run_aapt2_compile()
  run_aapt2_link()
  run_java_complie()
  convert_dex()
  add_to_apk()
  align_apk()
  sign_apk()
  update_manifest()
  generate_keystore()
  
"""

def run_aapt2_compile():
	run([tool['aapt2'], "compile", "--dir", project['res'], "-o", output['res']])

def run_aapt2_link():
	flat_files = find_files(output['res'], ".flat")

	run([
		tool['aapt2'], "link", "-o", 
		path(output['dir'], "temp.apk"), "-I", tool['platform'],
		"--manifest", output['manifest'], "--java", output['gen'],
		*flat_files
	])

def run_java_compile():
	java_files = find_files(path(Project_dir, "src"), ".java")
	jar_files = []
	if os.path.isdir(project['jar_lib']):
		jar_files = find_files(project['jar_lib'], ".jar")
		for jar in jar_files:
			VexelLog(f"Loaded {os.path.basename(jar)}", "INFO")
			
	gen_files = find_files(output['gen'], ".java")
	# Line [current + 2] uses a Linux/Android only feature this 
	# has to be changed later for window capability
	classpath = ":".join([tool['platform']] + jar_files)

	run([
		tool['javac'],
		"-d", output['classes'],
		"-classpath", classpath,
		*(java_files + gen_files)
	])

def convert_dex():
	class_files = find_files(output['classes'], ".class")
	if not class_files:
		raise CodeNotFoundError("Unable to find Class files")
	jar_files = []
	if os.path.isdir(project['jar_lib']):
		jar_files.extend(find_files(project['jar_lib'], ".jar"))
		
	run([
		tool['java'], "-cp", tool['d8'], "com.android.tools.r8.D8",
		"--lib", tool['platform'], "--min-api", "21", "--output", 
		output['dex'], *class_files, *jar_files
	])

def add_to_apk(apk_path, source_path, apk_inner_path="", not_exists_ok=True, zip_level=6):
	if not os.path.exists(source_path):
		if not_exists_ok: return
		raise FileNotFoundError(f"Path not found: {source_path}")

	temp_apk = apk_path + ".tmp"

	with zipfile.ZipFile(apk_path, "r") as zin:
		with zipfile.ZipFile(temp_apk, "w",
			compression=zipfile.ZIP_DEFLATED, compresslevel=zip_level
		) as zout:

			# Copy old APK files
			for item in zin.infolist():
				zout.writestr(item, zin.read(item.filename), compress_type=item.compress_type)
				
			if os.path.isfile(source_path):
				if apk_inner_path:
					apk_file_path = apk_inner_path
				else:
					apk_file_path = os.path.basename(source_path)

				with open(source_path, "rb") as f:
					zout.writestr(apk_file_path, f.read())

				VexelLog(f"Added {os.path.basename(apk_file_path)}", "INFO")
				
			elif os.path.isdir(source_path):
				for root, _, files in os.walk(source_path):
					for file in files:

						full_path = os.path.join(root, file)
						rel_path = os.path.relpath(full_path, source_path)
						if apk_inner_path:
							apk_file_path = path(apk_inner_path, rel_path).replace("\\", "/")
						else:
							apk_file_path = rel_path.replace("\\", "/")
						zout.write(full_path, apk_file_path)
						VexelLog(f"Added {apk_file_path}", "INFO")
	os.replace(temp_apk, apk_path)
	
def align_apk():
	run([
		tool['zipalign'], "-f", "4",
		path(output['dir'], "temp_unaligned.apk"),
		path(output['dir'], "Debug.apk")
	])

def sign_apk():
	keystore_path, password = generate_keystore(output['key'])
		
	run([
		tool['java'], "-jar", tool['apksigner'], "sign",
		"--ks", keystore_path,
		"--ks-pass", f"pass:{password}",
		path(output['dir'], "Debug.apk")
	])

def update_manifest(manifest_path, out_manifest_path, cfg):
	if not check_file(manifest_path):
		VexelLog("Failed to load AndroidManifest.xml", "CRASH")
		
	tree = ET.parse(manifest_path)
	root = tree.getroot()

	android_ns = "http://schemas.android.com/apk/res/android"
	ET.register_namespace('android', android_ns)

	def android_attr(name):
		return f"{{{android_ns}}}{name}"

	# ---- APP ---- #
	app = cfg.get("app", {})
	if app:
		if "package" in app:
			root.set("package", app["package"])

		if "version_code" in app:
			root.set(android_attr("versionCode"), str(app["version_code"]))

		if "version_name" in app:
			root.set(android_attr("versionName"), str(app["version_name"]))

		application = root.find("application")
		if application is not None and "name" in app:
			application.set(android_attr("label"), app["name"])

	# ---- SDK ---- #
	sdk = cfg.get("sdk", {})
	if sdk:
		uses_sdk = root.find("uses-sdk")
		if uses_sdk is None:
			uses_sdk = ET.SubElement(root, "uses-sdk")

		if "min" in sdk:
			uses_sdk.set(android_attr("minSdkVersion"), str(sdk["min"]))

		if "target" in sdk:
			uses_sdk.set(android_attr("targetSdkVersion"), str(sdk["target"]))

	tree.write(out_manifest_path, encoding="utf-8", xml_declaration=True)
	
def generate_keystore(Out_Key_dir, length=24):
	keystore_path = path(Out_Key_dir, "debug.keystore")

	if os.path.exists(keystore_path):
		with open(path(Out_Key_dir, "debug.keystore.pass")) as f:
			password = f.read().strip()
			return keystore_path, password

	VexelLog("Generating new keystore...", "INFO")

	chars = string.ascii_letters + string.digits
	password = ''.join(secrets.choice(chars) for _ in range(length))

	run([
		tool['keytool'],
		"-genkeypair",
		"-v",
		"-keystore", keystore_path,
		"-storepass", password,
		"-keypass", password,
		"-alias", "vexelkey",
		"-keyalg", "RSA",
		"-keysize", "2048",
		"-validity", "10000",
		"-dname", "CN=Vexel,O=Dev,C=US"
	])

	# Save password (IMPORTANT)
	with open(path(Out_Key_dir, "debug.keystore.pass"), "w") as f:
		f.write(password)
	VexelLog("Keystore created", "INFO")

	return keystore_path, password

File: src/BuildUtils.kt
// ============================================================
//						  BuildUtils.kt
// ============================================================
package app.pie.vexel

import java.io.File
import java.io.FileWriter
import java.io.FileOutputStream
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.io.FileNotFoundException

import java.security.SecureRandom

import net.lingala.zip4j.ZipFile
import net.lingala.zip4j.exception.ZipException
import net.lingala.zip4j.model.ZipParameters

import org.jdom2.Element
import org.jdom2.Namespace
import org.jdom2.input.SAXBuilder
import org.jdom2.output.Format
import org.jdom2.output.XMLOutputter

data class KeyStoreResult(
	val keystorePath: String,
	val password: String
)

object Build {
	// AAPT2
	fun runAapt2Compile() {
		BuildTools.runAapt2(listOf("compile", "--dir", Vexel.Project.res, "-o", Vexel.Output.res))
	}
	fun runAapt2Link() {
		val flatFiles = findFiles(Vexel.Output.res, ".flat")
		if (flatFiles.isEmpty()) { return }
		
		BuildTools.runAapt2(listOf("link",
			"-o", path(Vexel.Output.apk, "base.apk"),
			"-I", BuildTools.getSdk(),
			"--manifest", Vexel.Output.manifest,
			"--java", Vexel.Output.gen
			) + flatFiles
		)
	}
	
	// Java
	fun runJavaCompile() {
		val javaFiles = findFiles(path(Vexel.Project.dir, "java"), ".java")
		val genFiles = findFiles(Vexel.Output.gen, ".java")
		val jarFiles = mutableListOf<String>()
		
		if (checkDir(Vexel.Project.jarLib)) {
			jarFiles.addAll(findFiles(Vexel.Project.jarLib, ".jar"))
			for (jar in jarFiles) {
				vexelLog("Loaded ${File(jar).name}", "INFO")
			}
		}
		val separator = File.pathSeparator
		val classpath = (listOf(BuildTools.getSdk()) + jarFiles).joinToString(separator)
		
		BuildTools.runJavac(listOf(
			"-d", Vexel.Output.classes,
			"-classpath", classpath
			) + javaFiles + genFiles
		)
	}
	
	// Dex
	fun convertDex() {
		val classFiles = findFiles(Vexel.Output.classes, ".class")
		if (classFiles.isEmpty()) {
			vexelThrow(ErrorCode.FILE_NOT_FOUND,
			"Unable to find class files",
			"Verify that the java code exists and it is compiled successfully"
			)
		}
		
		val jarFiles = mutableListOf<String>()
		
		if (checkDir(Vexel.Project.jarLib)) {
			jarFiles.addAll(findFiles(Vexel.Project.jarLib, ".jar"))
		}
		
		BuildTools.runD8(listOf(
			"--lib", BuildTools.getSdk(),
			"--min-api", "21",
			"--output", Vexel.Output.dex
			) + classFiles + jarFiles
		)
	}
	
	// AddToZip
	fun addToZip(zipPath: String, filePath: String, innerPath: String = "", okIfMissing: Boolean = true) {
		val sourceFile = File(filePath)
		
		// Check source file
		if (!checkFile(filePath) && !checkDir(filePath)) {
			if (okIfMissing) { return }
			else {
				vexelThrow(ErrorCode.FILE_NOT_FOUND,
				"Failed to find source file: $filePath",
				"Check that the file path is correct and have permission to read it."
				)
			}
		}
		
		if (!File(zipPath).exists()) {
			vexelThrow(ErrorCode.FILE_NOT_FOUND,
			"Failed to find zip file: $zipPath",
			"Check that the file path is correct and have permission to read it."
			)
		}
		
		try {
			val zipFile = ZipFile(zipPath)
			if (sourceFile.isDirectory) {
				zipFile.addFolder(sourceFile)
			} else {
				zipFile.addFile(sourceFile)
			}
			
		} catch (e: ZipException) {
			vexelThrow(ErrorCode.ZIP_ERROR,
			"Failed to add file to zip",
			"Check that both zip and file exists.",
			debugInfo = "ZipPath: $zipPath \nFilePath: $filePath"
			)
		}
	}
	
	// Zipalign
	fun alignApk() {
		BuildTools.runZipalign(listOf("-f", "4",
			path(Vexel.Output.apk, "unaligned.apk"),
			path(Vexel.Output.apk, "final.apk")
			)
		)
	}
	
	// Generate KeyStore
	fun generateKeystore(outKeyDir: String, key: String? = null, length: Int = 24): KeyStoreResult {
		val keystorePath = path(outKeyDir, "debug.keystore")
		
		if (checkFile(path(outKeyDir, "debug.keystore.pass"))) {
			val file = File(path(outKeyDir, "debug.keystore.pass"))
			val key: String = file.readText() 
		}
		
		val password = if (key != null) { key } else {
			vexelLog("Generating new password...", "INFO")
			
			val chars = "abcdefghijklmnopqrstuvwxyz" + "ABCDEFGHIJKLMNOPQRSTUVWXYZ" + "0123456789"
			val random = SecureRandom()
			
			buildString {
				repeat(length) {
					append(chars[random.nextInt(chars.length)])
				}
			}
		}
		
		vexelLog("Generating keystore...", "INFO")
		
		BuildTools.runKeytool(listOf(
			"-genkeypair", "-v",
			"-keystore", keystorePath,
			"-storepass", password,
			"-keypass", password,
			"-alias", "vexelkey",
			"-keyalg", "RSA",
			"-keysize", "2048",
			"-validity", "10000",
			"-dname", "CN=Vexel,O=Dev,C=US"
			)
		)
		
		writeFile(path(outKeyDir, "debug.keystore.pass"), password)
		vexelLog("Keystore created", "INFO")
		
		return KeyStoreResult(keystorePath, password)
	}
	
	// ApkSigner
	fun signApk() {
		val result = generateKeystore(Vexel.Output.key)
		
		BuildTools.runApkSigner(listOf(
			"sign", "--ks", result.keystorePath,
			"--ks-pass", "pass:${result.password}",
			path(Vexel.Output.apk, "final.apk")
			)
		)
	}
	
	// Generate Manifest
	fun updateManifest(manifestPath: String, outManifestPath: String, cfg: Map<String, Any>) {
		if (!checkFile(manifestPath)) {
			vexelThrow(ErrorCode.FILE_NOT_FOUND,
			"Failed to load AndroidManifest.xml",
			"Check that AndroidManifest.xml exists."
			)
			return
		}

		val androidNs = Namespace.getNamespace("android", "http://schemas.android.com/apk/res/android")
	
		try {
			val document = SAXBuilder().build(File(manifestPath))
			val root = document.rootElement
		
			// ---------------- APP ----------------
			val app = cfg["app"] as? Map<*, *>
		
			app?.let {
				it["package"]?.let { value ->
				root.setAttribute("package", value.toString()) }

				it["version_code"]?.let { value ->
				root.setAttribute("versionCode", value.toString(), androidNs) }

				it["version_name"]?.let { value ->
				root.setAttribute("versionName", value.toString(), androidNs) }

				val application = root.getChild("application")

				if (application != null) {
					it["name"]?.let { value -> application.setAttribute("label", value.toString(), androidNs) }
				}
			}

			// ---------------- SDK ----------------
			val sdk = cfg["sdk"] as? Map<*, *>

			sdk?.let {
				val usesSdk = root.getChild("uses-sdk")
				?: Element("uses-sdk").also { root.addContent(it) }

				it["min"]?.let { value ->
				usesSdk.setAttribute("minSdkVersion", value.toString(), androidNs) }

				it["target"]?.let { value ->
				usesSdk.setAttribute("targetSdkVersion", value.toString(), androidNs) }
			}
			// ---------------- SAVE ----------------
			val outputter = XMLOutputter(
				Format.getPrettyFormat().apply {
					omitDeclaration = false
					encoding = "UTF-8"
				}
			)
			
			val outFile = File(outManifestPath)
			outFile.parentFile?.mkdirs()
		
			FileWriter(outFile).use { writer -> outputter.output(document, writer) }
			
		} catch (e: Exception) {
			vexelThrow(ErrorCode.INVALID_MANIFEST,
			"Failed to parse AndroidManifest.xml",
			"Verify that the manifest is valid XML and contains the required elements.",
			cause = e
			)
		}
	}
}